clients/lsp-rust.el

```diff
@@ -945,7 +946,9 @@ or JSON objects in `rust-project.json` format."
   :library-folders-fn (lambda (_workspace) lsp-rust-analyzer-library-directories)
   :after-open-fn (lambda ()
                    (when lsp-rust-analyzer-server-display-inlay-hints
-                     (lsp-rust-analyzer-inlay-hints-mode)))
+                     (lsp-rust-analyzer-inlay-hints-mode))
+                   (when lsp-semantic-tokens-enable
```

Comment:
Does this work for you ?

I tried this approach in terraform mode and I could see that the terraform language server was not getting initialized with the proper values.

Reply:
It does work for me for lsp-rust and rust-analyzer using my branch https://github.com/alanz/lsp-mode/tree/semantic-tokens-plus-optimise-modifiers which is a combination of https://github.com/emacs-lsp/lsp-mode/pull/3668 and https://github.com/emacs-lsp/lsp-mode/pull/3670

Reply:
I tried again and these are my observations:

- The buffer local variable is actually set properly.
- The terraform language server is not initialized with the right values. I believe the `after-open-fn` is too late in the lifecycle process to change the initialization values.

It could be that your other changes in your branch may cause this to work. I will wait till your changes are merged, before giving it an attempt again. 

I think a more proper fix for this would be to allow proper initialization in the `make-lsp-client` function itself.

Reply:
It might relate to the initial capability negotiation.  The rust language server ignores the capabilities [the client sends](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#semanticTokensClientCapabilities), and always returns a fixed modifiers legend (fixed for the given version of the language server).

Reply:
Yeah, that's probably it.